daffodil-runtime2/src/main/resources/c/libruntime/errors.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ERRORS_H
+#define ERRORS_H
+
+#include <stdio.h>    // for FILE, size_t
+#include <stdint.h>   // for int64_t
+
+// ErrorCode - types of errors which could occur
+
+enum ErrorCode
+{
+    ERR_CHOICE_KEY,
+    ERR_FILE_CLOSE,
+    ERR_FILE_FLUSH,
+    ERR_FILE_OPEN,
+    ERR_FIXED_VALUE,
+    ERR_INFOSET_READ,
+    ERR_INFOSET_WRITE,
+    ERR_PARSE_BOOL,
+    ERR_STACK_EMPTY,
+    ERR_STACK_OVERFLOW,
+    ERR_STACK_UNDERFLOW,
+    ERR_STREAM_EOF,
+    ERR_STREAM_ERROR,
+    ERR_STRTOBOOL,
+    ERR_STRTOD_ERRNO,
+    ERR_STRTOI_ERRNO,
+    ERR_STRTONUM_EMPTY,
+    ERR_STRTONUM_NOT,
+    ERR_STRTONUM_RANGE,
+    ERR_XML_DECL,
+    ERR_XML_ELEMENT,
+    ERR_XML_ERD,
+    ERR_XML_GONE,
+    ERR_XML_INPUT,
+    ERR_XML_LEFT,
+    ERR_XML_MISMATCH,
+    ERR_XML_WRITE
+};
+
+// Error - specific error occuring now
+
+typedef struct Error
+{
+    enum ErrorCode code;
+    union
+    {
+        const char *s;   // for %s
```

Comment:
Oh, I didn't consider passing schema filename/lineno information for a diagnostic.  We'd still want to pass a string or int64 for some errors, so I would store an ERD (or a Location object) in its own separate Error member field, not as an alternative field in the union.  We currently don't store any schema filename/lineno in the ERD metadata, so I'll add this idea to the runtime2 todos for a later time.  We also might want to pass a PState/UState for some errors to convey filename/position information about where the error was detected in the data, yes?

Reply:
Yes, in the regular scala backend, the parse error method is actually a method of the PState/UState object, so that's always implicitly passed. We depend on the fact that the state has all the static information accessible. In this runtime, perhaps we have to pass the static runtime-data object as well as the state object. 

In the standard scala backend, there are runtime-data objects that are not ERD, i.e., they are associated with choices, variables, etc. We should plan on a need for that also. More than just "elements" need runtime data containing static information.  

Basically, everything that could ever need runtime context (which is everything, since this is a runtime/backend) should be passed a state, and either the state provides access to the runtime data, or the runtime data (static context) should also be passed.